### PR TITLE
Fix: Make sure decks created in elastic task workers are transferred to parent process

### PR DESCRIPTION
**Pull request by fg91**
# TL;DR
Currently, Flyte decks don't work with the elastic tasks. This PR fixes this.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue <- no pending items

## Complete description
The reason is that the elastic task starts worker processes (as `torchrun` does). When a user creates a deck, this deck is appended to `flytekit.current_context().decks` *in the worker processes*. However, the upload of the deck happens based on the `ctx` in the parent process where these decks are never appended. The decks are, thus, never shown in the UI.

This PR fixes this by returning the list of decks from the worker processes and, for rank `0`, attaches those decks to the parent processes context.

## Tracking Issue
_NA_

## Follow-up issue
_NA_

### CODE REVIEW

**Review by wild-endeavor — DISMISSED** [review]

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py [hunk 1]
```diff
@@ -203,7 +203,22 @@ def get_custom(self, settings: SerializationSettings) -> Dict[str, Any]:
...
+    """
+    A named tuple representing the result of a torch elastic worker process.
+
+    Attributes:
+        return_value (Any): The value returned by the task function in the worker process.
+        decks (list[flytekit.Deck]): A list of flytekit Deck objects created in the worker process.
+    """
+
+    return_value: Any
+    decks: list[flytekit.Deck]
```
Comment (fg91) [inline plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:216]:
```suggestion
    decks: List[flytekit.Deck]
```

CI fails with:

```console
flytekitplugins/kfpytorch/task.py:216: in ElasticWorkerResult
    decks: list[flytekit.Deck]
E   TypeError: 'type' object is not subscriptable
```

Seems to use a python version that doesn't support this yet

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 import os
 from dataclasses import dataclass, field
 from enum import Enum
-from typing import Any, Callable, Dict, Optional, Union
+from typing import Any, Callable, Dict, NamedTuple, Optional, Union
```
Comment (fg91) [inline plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:8]:
```suggestion
from typing import Any, Callable, Dict, List, NamedTuple, Optional, Union
```

**Comment by fg91** [comment]
Thanks for the quick review @wild-endeavor ! 🙏 

I pushed two additional commits (https://github.com/flyteorg/flytekit/pull/1837/commits/027feb03af38fa4f70e03aee051a6b345542d7c4, https://github.com/flyteorg/flytekit/pull/1837/commits/7f6fd68bb58fdccaf2f2d2e254a2bac01727c149) as CI failed, assuming because a python version is used that doesn't yet support `decks: list[flytekit.Deck]` instead of `List[...`.

[Build / build-integration (ubuntu-latest, 3.11) (pull_request) ](https://github.com/flyteorg/flytekit/actions/runs/6225233283/job/16895207419?pr=1837) still fails with this though:

```
The push refers to repository [localhost:30000/flytekit]
Get "http://localhost:30000/v2/": read tcp 127.0.0.1:34148->127.0.0.1:30000: read: connection reset by peer
Error: Process completed with exit code 1.
```

Unrelated I assume?

**Review by pingsutw — APPROVED** [review]

**Comment by pingsutw** [comment]
That's unrelated. We've fixed it in another PR.
